cli/kata-check.go

```diff
@@ -293,3 +301,58 @@ var kataCheckCLICommand = cli.Command{
 		return nil
 	},
 }
+
+func archKernelParamHandler(onVMM bool, fields logrus.Fields, msg string) bool {
+        param, ok := fields["parameter"].(string)
+        if !ok {
+                return false
+        }
+
+        // This option is not required when
+        // already running under a hypervisor.
+        if param == "unrestricted_guest" && onVMM {
+                kataLog.WithFields(fields).Warn(kernelPropertyCorrect)
+                return true
+        }
+
+        if param == "nested" {
+                kataLog.WithFields(fields).Warn(msg)
+                return true
+        }
+
+        // don't ignore the error
+        return false
+}
+
+// genericKvmIsUsable determines if it will be possible to create a full virtual machine
+// by creating a minimal VM and then deleting it.
+func genericKvmIsUsable() error {
+        flags := syscall.O_RDWR | syscall.O_CLOEXEC
+
+        f, err := syscall.Open(kvmDevice, flags, 0)
+        if err != nil {
+                return err
+        }
+        defer syscall.Close(f)
+
+        fieldLogger := kataLog.WithField("check-type", "full")
+
+        fieldLogger.WithField("device", kvmDevice).Info("device available")
+
+        vm, _, errno := syscall.Syscall(syscall.SYS_IOCTL,
```

Comment:
Not related to this patch, but is there any possibility that we have `/dev/kvm` device but kvm is not enabled? I wonder if the extra step of checking kvm ioctl is necessary. If not, we can remove the only piece of C go from kata runtime.

Reply:
Actually, yes (kind of). This code was added to protect against the scenario where the user has a different hypervisor running (VMWare, VirtualBox, etc) since they stop qemu from using kvm. See:

- https://github.com/clearcontainers/runtime/issues/848

Reply:
That's also why `kata-check` is specified twice in the README, one with `sudo` and one without since `kata-check` needs `root` access to create a KVM VM (which we then destroy), but you can run the command without `root` (but we can't then detect if another hypervisor is using it):

- https://github.com/kata-containers/runtime#hardware-requirements

Reply:
OK, fair enough. I guess next possibility is to either push the kvm ioctl number to go runtime, or define it in govmm. Then we can still remove the c go code.